migrations/back-37/back-37.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/urfave/cli"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/tidepool-org/platform/application"
+	"github.com/tidepool-org/platform/data/blood/glucose"
+	"github.com/tidepool-org/platform/errors"
+	migrationMongo "github.com/tidepool-org/platform/migration/mongo"
+	storeStructuredMongo "github.com/tidepool-org/platform/store/structured/mongo"
+)
+
+func main() {
+	application.RunAndExit(NewMigration())
+}
+
+type Migration struct {
+	*migrationMongo.Migration
+	dataRepository *storeStructuredMongo.Repository
+}
+
+func NewMigration() *Migration {
+	return &Migration{
+		Migration: migrationMongo.NewMigration(),
+	}
+}
+
+func (m *Migration) Initialize(provider application.Provider) error {
+	if err := m.Migration.Initialize(provider); err != nil {
+		return err
+	}
+
+	m.CLI().Usage = "BACK-37: Migrate all existing data to add required Platform deduplication hash fields"
+	m.CLI().Description = "BACK-37: To fully migrate devices from the `jellyfish` upload API to the `platform` upload API"
+	m.CLI().Authors = []cli.Author{
+		{
+			Name:  "J H BATE",
+			Email: "jamie@tidepool.org",
+		},
+	}
+
+	m.CLI().Action = func(ctx *cli.Context) error {
+		if !m.ParseContext(ctx) {
+			return nil
+		}
+		return m.execute()
+	}
+
+	return nil
+}
+
+func (m *Migration) execute() error {
+	m.Logger().Debug("Migrate jellyfish upload API data")
+	m.Logger().Debug("Creating data store")
+
+	mongoConfig := m.NewMongoConfig()
+	mongoConfig.Database = "data"
+	mongoConfig.Timeout = 60 * time.Minute
+	dataStore, err := storeStructuredMongo.NewStore(mongoConfig)
+	if err != nil {
+		return errors.Wrap(err, "unable to create data store")
+	}
+	defer dataStore.Terminate(context.Background())
+
+	m.Logger().Debug("Creating data repository")
+
+	m.dataRepository = dataStore.GetRepository("deviceData")
+	hashUpdatedCount, archivedCount, errorCount := m.migrateJellyfishDocuments()
+	m.Logger().Infof("Migrated %d jellyfish documents", hashUpdatedCount)
+	m.Logger().Infof("Archived %d duplicate jellyfish documents", archivedCount)
+	m.Logger().Infof("%d errors occurred", errorCount)
+
+	return nil
+}
+
+func (m *Migration) migrateJellyfishDocuments() (int, int, int) {
+	logger := m.Logger()
+
+	logger.Debug("Finding distinct users")
+
+	var hashUpdatedCount, archivedCount, errorCount int
+
+	userIDs, err := m.dataRepository.Distinct(context.Background(), "_userId", bson.M{})
```

Comment:
I don't remember, is there an index on `_userId`? I'm pretty sure there is. If not, this would have to do a scan through all records.

Reply:
Also, consider adding the same selector as below so we are only getting distinct user ids for those users with the selector. Specifically:
```
			selector := bson.M{
				"_active": true,
				// uploads aren't de-duped.
				"type":          bson.M{"$ne": "upload"},
				"_deduplicator": bson.M{"$exists": false},
			}
```

Reply:
Ignore for now.

Reply:
Consider total number of distinct users in memory at the same time.